syft/pkg/cataloger/golang/licenses.go

```diff
@@ -0,0 +1,108 @@
+package golang
+
+import (
+	"fmt"
+	"os"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/mitchellh/go-homedir"
+
+	"github.com/anchore/syft/internal/licenses"
+	"github.com/anchore/syft/internal/log"
+	"github.com/anchore/syft/syft/source"
+)
+
+type goLicenses struct {
+	searchLocalModCacheLicenses bool
+	localModCacheResolver       source.FileResolver
+}
+
+func newGoLicenses(searchLocalModCacheLicenses bool) goLicenses {
+	return goLicenses{
+		searchLocalModCacheLicenses: searchLocalModCacheLicenses,
+		localModCacheResolver:       deferredModCacheResolver,
+	}
+}
+
+// this needs to be shared between GoMod & GoBinary so it's only scanned once
+var deferredModCacheResolver = newDeferredModCacheResolver()
+
+func newDeferredModCacheResolver() source.FileResolver {
+	return source.NewDeferredResolverFromSource(func() (source.Source, error) {
+		goPath := os.Getenv("GOPATH")
+
+		if goPath == "" {
+			homeDir, err := homedir.Dir()
+			if err != nil {
+				log.Debug("unable to determine user home dir: %v", err)
+			}
+			goPath = path.Join(homeDir, "go")
```

Comment:
it doesn't look like we should use the value of `homeDir` if error != nil... should we return an error instead (and remove the log statement)?

Reply:
See above comment. I don't think we should try and recreate the `GOPATH`. Either it is set and we look there, or it is not, and we do not even try.

If we get requests later that _lots of people_ want us to determine this automatically, we always can add it later.

Reply:
actually, I think I like that better. If `goPath` is empty, then return don't return a resolver to search against at all.

Reply:
According [to the go docs](https://go.dev/doc/gopath_code), the default gopath is:

> The GOPATH environment variable specifies the location of your workspace. It defaults to a directory named go inside your home directory, so $HOME/go on Unix, $home/go on Plan 9, and %USERPROFILE%\go (usually C:\Users\YourName\go) on Windows.

Why wouldn't we default to this same location if GOPATH is not explicitly set?